### PR TITLE
GLTexture.Texture constructor with image parameter of type LinearAlgebra.Adjoint

### LINKED ISSUE #1320: GLMakie: heatmap errors on adjoint(::Matrix{Float32})
Environment: Julia v1.6.2, Makie v0.15.2, GLMakie v0.4.6.

MWE:

```
julia> d = rand(Float32, 10, 5)
10×5 Matrix{Float32}:
 0.146433   0.665632   0.595443  0.101326   0.459009
 0.826909   0.212982   0.404034  0.207223   0.274588
 0.915337   0.320892   0.721542  0.14295    0.89392
 0.457666   0.888185   0.830775  0.251421   0.223386
 0.125917   0.259368   0.130264  0.770156   0.288838
 0.0780659  0.886574   0.231801  0.0552516  0.497002
 0.809111   0.105489   0.955071  0.0478306  0.428939
 0.756231   0.0787696  0.941425  0.353625   0.527136
 0.367141   0.183711   0.113805  0.378282   0.962215
 0.84761    0.628935   0.705939  0.170442   0.52278

# This works:
julia> heatmap(d)

# But this doesn't:
julia> heatmap(d')
Error showing value of type Makie.FigureAxisPlot:
ERROR: MethodError: no method matching GLMakie.GLAbstraction.Texture(::LinearAlgebra.Adjoint{Float32, Matrix{Float32}}; minfilter=:nearest)
Closest candidates are:
  (::Type{T})(::Observable; kw...) where T<:GLMakie.GLAbstraction.GPUArray at /home/fotech/.julia/packages/GLMakie/TViqo/src/GLAbstraction/AbstractGPUArray.jl:209
  GLMakie.GLAbstraction.Texture(::Array{Matrix{T}, 1}; internalformat, texturetype, format, parameters...) where T<:Union{Real, ColorTypes.Colorant, StaticArrays.StaticVector{N, T} where {N, T}} at /home/fotech/.julia/packages/GLMakie/TViqo/src/GLAbstraction/GLTexture.jl:131
  GLMakie.GLAbstraction.Texture(::ShaderAbstractions.Sampler{T, N, Data} where Data; kwargs...) where {T, N} at /home/fotech/.julia/packages/GLMakie/TViqo/src/GLAbstraction/GLTexture.jl:117
  ...
Stacktrace:
  [1] GLMakie.GLAbstraction.Texture(x::Observable{LinearAlgebra.Adjoint{Float32, Matrix{Float32}}}; kw::Base.Iterators.Pairs{Symbol, Symbol, Tuple{Symbol}, NamedTuple{(:minfilter,), Tuple{Symbol}}})
    @ GLMakie.GLAbstraction ~/.julia/packages/GLMakie/TViqo/src/GLAbstraction/AbstractGPUArray.jl:210
  [2] (::GLMakie.var"#130#135"{Scene, Heatmap{Tuple{IntervalSets.ClosedInterval{Float32}, IntervalSets.ClosedInterval{Float32}, LinearAlgebra.Adjoint{Float32, Matrix{Float32}}}}})(gl_attributes::Dict{Symbol, Any})
    @ GLMakie ~/.julia/packages/GLMakie/TViqo/src/drawing_primitives.jl:393
  [3] (::GLMakie.var"#74#80"{GLMakie.var"#130#135"{Scene, Heatmap{Tuple{IntervalSets.ClosedInterval{Float32}, IntervalSets.ClosedInterval{Float32}, LinearAlgebra.Adjoint{Float32, Matrix{Float32}}}}}, GLMakie.Screen, Scene, Heatmap{Tuple{IntervalSets.ClosedInterval{Float32}, IntervalSets.ClosedInterval{Float32}, LinearAlgebra.Adjoint{Float32, Matrix{Float32}}}}})()
    @ GLMakie ~/.julia/packages/GLMakie/TViqo/src/drawing_primitives.jl:75
  [4] get!(default::GLMakie.var"#74#80"{GLMakie.var"#130#135"{Scene, Heatmap{Tuple{IntervalSets.ClosedInterval{Float32}, IntervalSets.ClosedInterval{Float32}, LinearAlgebra.Adjoint{Float32, Matrix{Float32}}}}}, GLMakie.Screen, Scene, Heatmap{Tuple{IntervalSets.ClosedInterval{Float32}, IntervalSets.ClosedInterval{Float32}, LinearAlgebra.Adjoint{Float32, Matrix{Float32}}}}}, h::Dict{UInt64, GLMakie.GLAbstraction.RenderObject}, key::UInt64)
    @ Base ./dict.jl:465
  [5] cached_robj!(robj_func::GLMakie.var"#130#135"{Scene, Heatmap{Tuple{IntervalSets.ClosedInterval{Float32}, IntervalSets.ClosedInterval{Float32}, LinearAlgebra.Adjoint{Float32, Matrix{Float32}}}}}, screen::GLMakie.Screen, scene::Scene, x::Heatmap{Tuple{IntervalSets.ClosedInterval{Float32}, IntervalSets.ClosedInterval{Float32}, LinearAlgebra.Adjoint{Float32, Matrix{Float32}}}})
    @ GLMakie ~/.julia/packages/GLMakie/TViqo/src/drawing_primitives.jl:51
  [6] draw_atomic(screen::GLMakie.Screen, scene::Scene, x::Heatmap{Tuple{IntervalSets.ClosedInterval{Float32}, IntervalSets.ClosedInterval{Float32}, LinearAlgebra.Adjoint{Float32, Matrix{Float32}}}})
    @ GLMakie ~/.julia/packages/GLMakie/TViqo/src/drawing_primitives.jl:361
  [7] insert!(screen::GLMakie.Screen, scene::Scene, x::Combined)
    @ GLMakie ~/.julia/packages/GLMakie/TViqo/src/drawing_primitives.jl:176
  [8] insertplots!(screen::GLMakie.Screen, scene::Scene)
    @ GLMakie ~/.julia/packages/GLMakie/TViqo/src/screen.jl:56
  [9] #34
    @ ~/.julia/packages/GLMakie/TViqo/src/screen.jl:58 [inlined]
 [10] foreach(f::GLMakie.var"#34#36"{GLMakie.Screen}, itr::Vector{Scene})
    @ Base ./abstractarray.jl:2141
 [11] insertplots!(screen::GLMakie.Screen, scene::Scene)
    @ GLMakie ~/.julia/packages/GLMakie/TViqo/src/screen.jl:58
 [12] backend_display(screen::GLMakie.Screen, scene::Scene)
    @ GLMakie ~/.julia/packages/GLMakie/TViqo/src/display.jl:18
 [13] backend_display
    @ ~/.julia/packages/GLMakie/TViqo/src/display.jl:4 [inlined]
 [14] display(scene::Scene; update::Bool)
    @ Makie ~/.julia/packages/Makie/PFSZS/src/display.jl:60
 [15] display
    @ ~/.julia/packages/Makie/PFSZS/src/display.jl:56 [inlined]
 [16] #display#902
    @ ~/.julia/packages/Makie/PFSZS/src/display.jl:52 [inlined]
 [17] display
    @ ~/.julia/packages/Makie/PFSZS/src/display.jl:52 [inlined]
 [18] #display#901
    @ ~/.julia/packages/Makie/PFSZS/src/display.jl:51 [inlined]
 [19] display(fap::Makie.FigureAxisPlot)
    @ Makie ~/.julia/packages/Makie/PFSZS/src/display.jl:51
 [20] #invokelatest#2
    @ ./essentials.jl:708 [inlined]
 [21] invokelatest
    @ ./essentials.jl:706 [inlined]
 [22] print_response(errio::IO, response::Any, show_value::Bool, have_color::Bool, specialdisplay::Union{Nothing, AbstractDisplay})
    @ REPL /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/REPL/src/REPL.jl:247
 [23] (::REPL.var"#40#41"{REPL.LineEditREPL, Pair{Any, Bool}, Bool, Bool})(io::Any)
    @ REPL /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/REPL/src/REPL.jl:231
 [24] with_repl_linfo(f::Any, repl::REPL.LineEditREPL)
    @ REPL /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/REPL/src/REPL.jl:462
 [25] print_response(repl::REPL.AbstractREPL, response::Any, show_value::Bool, have_color::Bool)
    @ REPL /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/REPL/src/REPL.jl:229
 [26] (::REPL.var"#do_respond#61"{Bool, Bool, REPL.var"#72#82"{REPL.LineEditREPL, REPL.REPLHistoryProvider}, REPL.LineEditREPL, REPL.LineEdit.Prompt})(s::REPL.LineEdit.MIState, buf::Any, ok::Bool)
    @ REPL /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/REPL/src/REPL.jl:798
 [27] #invokelatest#2
    @ ./essentials.jl:708 [inlined]
 [28] invokelatest
    @ ./essentials.jl:706 [inlined]
 [29] run_interface(terminal::REPL.Terminals.TextTerminal, m::REPL.LineEdit.ModalInterface, s::REPL.LineEdit.MIState)
    @ REPL.LineEdit /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/REPL/src/LineEdit.jl:2441
 [30] run_frontend(repl::REPL.LineEditREPL, backend::REPL.REPLBackendRef)
    @ REPL /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/REPL/src/REPL.jl:1126
 [31] (::REPL.var"#44#49"{REPL.LineEditREPL, REPL.REPLBackendRef})()
    @ REPL ./task.jl:411

```

### LINKED ISSUE #1795: image fails with transpose of Matrix{Float32}
See the following MWE:

```
using Makie
a = zeros(Float32, 10, 10)
image(a')


Error showing value of type Makie.FigureAxisPlot:
ERROR: MethodError: no method matching GLMakie.GLAbstraction.Texture(::LinearAlgebra.Adjoint{Float32, Matrix{Float32}}; minfilter=:linear)
Closest candidates are:
  (::Type{T})(::Observable; kw...) where T<:GLMakie.GLAbstraction.GPUArray at ~/.julia/packages/GLMakie/TfSB1/src/GLAbstraction/AbstractGPUArray.jl:209
  GLMakie.GLAbstraction.Texture(::Array{Matrix{T}, 1}; internalformat, texturetype, format, parameters...) where T<:Union{Real, ColorTypes.Colorant, StaticVector} at ~/.julia/packages/GLMakie/TfSB1/src/GLAbstraction/GLTexture.jl:131
  GLMakie.GLAbstraction.Texture(::ShaderAbstractions.Sampler{T, N}; kwargs...) where {T, N} at ~/.julia/packages/GLMakie/TfSB1/src/GLAbstraction/GLTexture.jl:117
  ...
Stacktrace:
  [1] GLMakie.GLAbstraction.Texture(x::Observable{LinearAlgebra.Adjoint{Float32, Matrix{Float32}}}; kw::Base.Pairs{Symbol, Symbol, Tuple{Symbol}, NamedTuple{(:minfilter,), Tuple{Symbol}}})
    @ GLMakie.GLAbstraction ~/.julia/packages/GLMakie/TfSB1/src/GLAbstraction/AbstractGPUArray.jl:210
  [2] mesh_inner(mesh::Observable{GeometryBasics.Mesh{2, Float32, GeometryBasics.TriangleP{2, Float32, GeometryBasics.PointWithUV{2, Float32}}, GeometryBasics.FaceView{GeometryBasics.TriangleP{2, Float32, GeometryBasics.PointWithUV{2, Float32}}, GeometryBasics.PointWithUV{2, Float32}, GeometryBasics.NgonFace{3, GeometryBasics.OffsetInteger{-1, UInt32}}, StructArrays.StructVector{GeometryBasics.PointWithUV{2, Float32}, NamedTuple{(:position, :uv), Tuple{Vector{Point{2, Float32}}, Vector{Vec{2, Float32}}}}, Int64}, Vector{GeometryBasics.NgonFace{3, GeometryBasics.OffsetInteger{-1, UInt32}}}}}}, transfunc::Observable{Any}, gl_attributes::Dict{Symbol, Any})
    @ GLMakie ~/.julia/packages/GLMakie/TfSB1/src/drawing_primitives.jl:455
  [3] (::GLMakie.var"#151#154"{Image{Tuple{StepRangeLen{Float32, Float64, Float64, Int64}, StepRangeLen{Float32, Float64, Float64, Int64}, LinearAlgebra.Adjoint{Float32, Matrix{Float32}}}}})(gl_attributes::Dict{Symbol, Any})
    @ GLMakie ~/.julia/packages/GLMakie/TfSB1/src/drawing_primitives.jl:425
  [4] (::GLMakie.var"#87#92"{GLMakie.var"#151#154"{Image{Tuple{StepRangeLen{Float32, Float64, Float64, Int64}, StepRangeLen{Float32, Float64, Float64, Int64}, LinearAlgebra.Adjoint{Float32, Matrix{Float32}}}}}, GLMakie.Screen, Scene, Image{Tuple{StepRangeLen{Float32, Float64, Float64, Int64}, StepRangeLen{Float32, Float64, Float64, Int64}, LinearAlgebra.Adjoint{Float32, Matrix{Float32}}}}})()
    @ GLMakie ~/.julia/packages/GLMakie/TfSB1/src/drawing_primitives.jl:80
  [5] get!(default::GLMakie.var"#87#92"{GLMakie.var"#151#154"{Image{Tuple{StepRangeLen{Float32, Float64, Float64, Int64}, StepRangeLen{Float32, Float64, Float64, Int64}, LinearAlgebra.Adjoint{Float32, Matrix{Float32}}}}}, GLMakie.Screen, Scene, Image{Tuple{StepRangeLen{Float32, Float64, Float64, Int64}, StepRangeLen{Float32, Float64, Float64, Int64}, LinearAlgebra.Adjoint{Float32, Matrix{Float32}}}}}, h::Dict{UInt64, GLMakie.GLAbstraction.RenderObject}, key::UInt64)
    @ Base ./dict.jl:464
  [6] cached_robj!(robj_func::GLMakie.var"#151#154"{Image{Tuple{StepRangeLen{Float32, Float64, Float64, Int64}, StepRangeLen{Float32, Float64, Float64, Int64}, LinearAlgebra.Adjoint{Float32, Matrix{Float32}}}}}, screen::GLMakie.Screen, scene::Scene, x::Image{Tuple{StepRangeLen{Float32, Float64, Float64, Int64}, StepRangeLen{Float32, Float64, Float64, Int64}, LinearAlgebra.Adjoint{Float32, Matrix{Float32}}}})
    @ GLMakie ~/.julia/packages/GLMakie/TfSB1/src/drawing_primitives.jl:52
  [7] draw_atomic(screen::GLMakie.Screen, scene::Scene, x::Image{Tuple{StepRangeLen{Float32, Float64, Float64, Int64}, StepRangeLen{Float32, Float64, Float64, Int64}, LinearAlgebra.Adjoint{Float32, Matrix{Float32}}}})
    @ GLMakie ~/.julia/packages/GLMakie/TfSB1/src/drawing_primitives.jl:410
  [8] insert!(screen::GLMakie.Screen, scene::Scene, x::Combined)
    @ GLMakie ~/.julia/packages/GLMakie/TfSB1/src/drawing_primitives.jl:106
  [9] insertplots!(screen::GLMakie.Screen, scene::Scene)
    @ GLMakie ~/.julia/packages/GLMakie/TfSB1/src/screen.jl:56
 [10] #40
    @ ~/.julia/packages/GLMakie/TfSB1/src/screen.jl:58 [inlined]
 [11] foreach(f::GLMakie.var"#40#42"{GLMakie.Screen}, itr::Vector{Scene})
    @ Base ./abstractarray.jl:2694
 [12] insertplots!(screen::GLMakie.Screen, scene::Scene)
    @ GLMakie ~/.julia/packages/GLMakie/TfSB1/src/screen.jl:58
 [13] backend_display(screen::GLMakie.Screen, scene::Scene)
    @ GLMakie ~/.julia/packages/GLMakie/TfSB1/src/display.jl:18
 [14] backend_display
    @ ~/.julia/packages/GLMakie/TfSB1/src/display.jl:4 [inlined]
 [15] display(scene::Scene)
    @ Makie ~/.julia/packages/Makie/lgPZh/src/display.jl:59
 [16] #display#917
    @ ~/.julia/packages/Makie/lgPZh/src/display.jl:53 [inlined]
 [17] display
    @ ~/.julia/packages/Makie/lgPZh/src/display.jl:53 [inlined]
```

This can be solved by collecting the transpose but it was working before.

### PR DESCRIPTION
**Pull request by mfandl**
# Description

Fixes #1320
Fixes #1795

The array constructor for `GLMakie.GLAbstraction.Texture` had a concrete `Array{T, NDim}` type which caused calls to `Texture(...)` with transposed `Float32` arrays to result in an error (see linked issues for the actual message and stacktrace). This is because `LinearAlgebra.Adjoint{Float32, Matrix{Float32}} <: Array` does not hold.

This change fixes an issue when trying to produce `heatmap` or `image` from `Float32` arrays. Using different precision floats worked fine because these are converted to 32 bit floats such as in `tex = Texture(el32convert(mat), minfilter = interp)` and the adjoint is dropped.

I ran tests, however, I do not know whether I updated all the necessary places, or whether I did not cause a performance regression by this or something. Any pointers at improvements are welcome. The comment above the changed line mentions abstract arrays as types that are expected to work so I supposed this was a mistake?

MWE of the issue:

```julia
using GLMakie
heatmap(rand(Float32, 2, 3)')
```

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

## Checklist

- [ ] Added an entry in NEWS.md (for new features and breaking changes)
- [ ] Added or changed relevant sections in the documentation
- [ ] Added unit tests for new algorithms, conversion methods, etc.
- [ ] Added reference image tests for new plotting functions, recipes, visual options, etc.

### CODE REVIEW

File: GLMakie/src/GLAbstraction/GLTexture.jl [hunk 1]
```diff
@@ -134,7 +134,7 @@ So Array{Real, 2} == Texture2D with 1D Colorant dimension
 Array{Vec1/2/3/4, 2} == Texture2D with 1/2/3/4D Colorant dimension
 Colors from Colors.jl should mostly work as well
 """
-Texture(image::Array{T, NDim}; kw_args...) where {T <: GLArrayEltypes, NDim} =
+Texture(image::AbstractArray{T, NDim}; kw_args...) where {T <: GLArrayEltypes, NDim} =
```
Comment (SimonDanisch) [inline GLMakie/src/GLAbstraction/GLTexture.jl:137]: That's not correct...
`pointer(...)` is not guaranteed to be correctly implemented for all AbstractArrays, and might even return a garbage pointer if used on the wrong array type. A lot of AbstractArrays like view + adjoint even have pointer defined, which just point to the parent array, which is of course not what you want from a view/mapping.

You'll need something like `Texture(image::AbstractArray{T, NDim}) = Texture(convert(Array{T, NDIM}, image))` to be on the safe side.
I'm guessing this "fixed" your issue but the data must have been incorrectly permuted:
image

**Comment by SimonDanisch** [comment]
Thank you! As pointed out in the review, this was not a mistake, the "abstractarray" refers to static arrays like `Vec2f` for the element type.
Can you give an MWE, which currently fails sicne it isn't running into `el32convert`?

**Comment by mfandl** [comment]
@SimonDanisch Ah, thanks for the comments, I see why this is problematic now. Sorry for assuming that this was a mistake, I was just thinking that based on the comment in code.

> Can you give an MWE, which currently fails sicne it isn't running into `el32convert`?

I think the problem is not that `el32convert` wouldn't be called. It just acts as an identity function in the following case and therefore retains the currently unsupported type:

```julia
julia> x = rand(Float32, 2, 3)'
3×2 adjoint(::Matrix{Float32}) with eltype Float32:
 0.926551  0.0612232
 0.560767  0.139857
 0.617872  0.161374

julia> typeof(x)
LinearAlgebra.Adjoint{Float32, Matrix{Float32}}

julia> typeof(GLMakie.el32convert(x))
LinearAlgebra.Adjoint{Float32, Matrix{Float32}}
```

this is (understandably) different in case the element type is different:

```julia
julia> typeof(GLMakie.el32convert(rand(Float64, 2, 3)))
Matrix{Float32} (alias for Array{Float32, 2})
```

I thought this is actually correct behavior, that is why I considered updating the defintion to contain the `AbstractArray` type a proper solution.

Could you please help me understand what would be the the right direction? A solution along the lines of what you write above, or updating the behavior of el32 convert, or something else?

Anyway, the MWE of the problem I am trying to fix is this:

```julia
using GLMakie
heatmap(rand(Float32, 2, 3)')
```

I will also add it to the description of the PR.

**Comment by mfandl** [comment]
I updated the commit taking the code review comments into account - I reverted the `AbstractArray` change and introduced a new constructor which accepts the type which caused problems and collects it.

File: GLMakie/src/GLAbstraction/GLTexture.jl [hunk 2]
```diff
@@ -137,6 +137,9 @@ Colors from Colors.jl should mostly work as well
 Texture(image::Array{T, NDim}; kw_args...) where {T <: GLArrayEltypes, NDim} =
     Texture(pointer(image), size(image); kw_args...)::Texture{T, NDim}
 
+Texture(image::Adjoint{T, Array{T, NDim}}; kw_args...) where {T <: GLArrayEltypes, NDim} =
```
Comment (asinghvi17) [inline GLMakie/src/GLAbstraction/GLTexture.jl:140]: should this be `image::AbstractArray{T, NDim}` for the most generality?

Comment (mfandl) [reply GLMakie/src/GLAbstraction/GLTexture.jl:140]: Thank you for your review.

Originally this PR updated the `image::Array{T, NDim}` constructor to exactly what you write. But @SimonDanisch shed light on why this is wrong in his code review above. Would keeping the old `image::Array{T, NDim}` and adding an extra `image::AbstractArray{T, NDim}` constructor where I'd `collect` whatever the concrete `AbstractArray` subtype is make things safe? (Actually, probably not `collect` in the more general case, but the proposal from the first review: `Texture(image::AbstractArray{T, NDim}) = Texture(convert(Array{T, NDIM}, image))` - but the question remains, is this safe to do for any `AbstractArray`?)

You are of course right that more could be fixed in this PR now that I think about it more, this is another example that throws error of the same kind:

```julia
using GLMakie
x = rand(Float32, 5, 5);
y = @view x[2:3, 2:3];
heatmap(y)
```
```
Error showing value of type Makie.FigureAxisPlot:
ERROR: MethodError: no method matching GLMakie.GLAbstraction.Texture(::SubArray{Float32, 2, Matrix{Float32}, Tuple{UnitRange{Int64}, UnitRange{Int64}}, false}; minfilter=:nearest)
```
And the result of joining these two types really is an `AbstractArray`:

```
julia> typejoin(SubArray, Adjoint)
AbstractArray
```

**Pull request merged**